draft-zip.md

```diff
@@ -0,0 +1,109 @@
+4. **Future-Proofing the Network**: Diverting transaction fees into the `ZSF_BALANCE_AFTER[h]` is a forward-looking approach that prepares the Zcash network for future challenges and opportunities. It establishes a financial buffer that can be instrumental in addressing unforeseen issues and seizing strategic advantages as the Zcash ecosystem evolves.
+
+# Specification
+
+This ZIP only proposes a single modification to the transaction fees:
+1. Keep the current destination of 40% of the fees untouched, but route 60% of the fees back to `ZSF_BALANCE_AFTER[h]`
+
+## Transaction fee routing requirements
+
+- For each transaction, 60% of the total fee MUST be paid to the ZSF
+- Any fractions are rounded in favour of the miner _TODO ZIP owners: decide if you want rounding to favour the ZSF here_
+
+### Consensus Rule Changes
+
+The coinbase transaction at block height `height` MUST have a  `zsfDeposit(height)` that is greater than or equal to `floor(TransactionFees(height)) / 2)`, where `TransactionFees(height)` is the sum of the the remaining value in the transparent transaction value pool of the non-coinbase transactions in the block at `height`.
+
+_TODO ZIP owners: if you want rounding to favour the ZSF, use ceiling here_
+
+TODO ZIP Editors:
+- work out how to deal with pre-v6 transactions which don't have the zsfDeposit() field. For example, by requiring the remaining value in the transparent transaction value pool of a coinbase transaction to be greater than or equal to 60% of the fee
+- consider imposing this requirement on v6 transactions instead of an explicit `ZSF_DEPOSIT` field requirement
+
+# Rationale
+
+We believe that is ultimately a very minor change to the protocol, and quite simple in terms of implementation overhead. Additionally – and at the time of this writing – transaction fees are so small that 60% will likely not have a major impact.
+
+If transaction fees were to increase, future ZIPs can be written to change the 60%/40% split. Finding the optimal fee split may require an iterative approach involving adjustments based on real-world data and network dynamics.
+
+In the future, other ZIPs may be written to fund the ZSF in various ways, including but not limited to:
+- ZSA fees
+- dApp-specific fees and donations
+- “Storage fees” for any future data availability
+- Cross-chain bridge usage / Cross-chain messaging
+- Note sorting micro-transactional fees
+
+## Future Implications
+
+Looking into the future, there may come a time when the transaction fees become greater than the block reward issuance. At that time we may need to reconsider the 60/40 split. However, this will likely not be the case for the next 8-10 years due to forcasts based on issuance models and network traffic.
```

Comment:
There needs to be an explanation of the security goal, why reducing miner fees is secure now, and when it might not be secure in future.

Here is a draft security argument, but I'd like @daira to review it before it gets put into the ZIP:

# Rationale

## Mining Security

For Zcash to be secure, miners need financial incentives to:
* mine blocks,
* include transactions in those blocks, and
* choose to mine those blocks on the chain with the most work.

If miners don't have those incentives, then they might stop mining transactions, or deliberately create chain forks to re-mine transactions for themselves. These forks can roll back previously mined blocks and transactions, making Zcash transactions less secure and reliable.

Since the block reward is currently large in comparison to transaction fees, and the work required to create a block is significant, there is little incentive for miners to roll back blocks to take those transaction fees for themselves.

However, the incentive to mine a single transaction with a small fee is also small. So miners won't lose much by skipping mining a single transaction. This is a potential security issue, which can be mitigated by having multiple mining pools using diverse transaction selection policies.

Currently, miners only gain significant fees from mining large numbers of transactions (or logical actions) on the chain with the most work. This incentivises them to mine more transactions, because the amount of work required to mine a block is almost the same, regardless of the number of transactions.

Since miner fees are small, putting 60% of them into the ZSF does not significantly change the number of blocks a miner needs to mine to make the same profit. And the increase in issuance due to these ZSF deposits is also small.

## Future Security Considerations

If the block reward becomes small in comparison to miner fees, then Zcash security rests on:
* the minimum ZSF deposit fraction,
* the relative size of the block reward compared to median transaction fees,
* the relative size of the block reward compared to 90th percentile transaction fees, and
* the amount of work required to create a block.

However, smaller block rewards also make single transactions more valuable, so miners have less incentive to skip single transactions.

Here is an analysis of each of these factors:

If the minimum ZSF deposit fraction is small, then miners have a larger incentive to rollback and take transaction fees, rather than continuously mining to gain the deposit fees through later ZSF issuance.

For example, if the deposit fraction is 50%, then miners with less than 100% mining pool share have an incentive to rollback and take 50% of the transaction fee now, rather than waiting for later issuance of the ZSF deposit 50% over the long term. The break-even point depends on the timeframe. Every 4 years 50% of the ZSF is issued, so the break-even point over 4 years is around 33% miner fees, 33% ZSF fees re-issued, and 33% ZSF fees used for future issuance.

If median transaction fees are small, then miners have a smaller incentive to continuously mine many transactions. So a smaller ZSF deposit fraction would be more secure.

If the 90th percentile transactions fees are large, then miners have a larger incentive to rollback and take transaction fees. So a larger ZSF deposit fraction would be more secure, to incentivise waiting for future issuance.

If the amount of work required to create a block is small, then miners have a larger incentive to rollback blocks and try mining the same transactions themselves.

Reply:
Ty @daira and @teor2345 🙏

Reply:
@teor2345 's text here makes sense to me.

Oh geez, I started down a tangential exploration, and didn't stop until I got it all out. Apologies for the huge brain dump.

**tl/dr**: if fees were always equal to ZSF issuance (the "Steady-State Fee Rate" regime), the incentive to do the "Yoink"-style rollback attacks to capture transactions is always less than the zero-issuance regime.

(Now that I write that, it sounds incredibly obvious, but never-the-less here's a huge post analyzing this in detail.)

# Steady-State Fee Rate

I want to bring up a tangential exploration:

Let's imagine for simplicity, every block contains transactions that pay a constant $X$ of fees, and that a proportion $p$ are paid to the miner while $1-p$ are deposited into ZSF. If $(1-p)X$ = `BlockSubsidy(h)` (defined in #706), then the ZSF balance is unchanged in that block, because the same amount is deposited and withdrawn.

As long as the fee rate is a constant $X$ per block, _the ZSF balance cannot decrease_, so this scenario seems like a useful reference point to contrast with the scenario that "issuance is 0". Either scenarios seem potentially possible, though we should expect to see a scenario similar to SSFR for some period before the 0 issuance era.

In this scenario, the block miner receives $pX$ from direct fees, and `BlockSubsidy(h)` from ZSF disbursement, and the latter equals `pX`, so the miner receives $pX + (1-p)X = X$.

Let's call this the *Steady-State Fee Rate* (SSFR). If we set aside all security implications, the miner receives exactly the same revenue as in a Bitcoin-style blockchain where new coin issuance is 0, which I'll call *Bitcoin Post-Issuance* (BPI).

## Steady-State Fee Rate Security Brainstorm

Now let's (loosely) compare the security of SSFR & BPI with a constant fee rate of $X$:

### Txn Yoink Rollback Attack (Single Block), Constant $X$

In this attack, every block still has exactly $X$ fees, and a miner rolls back one block, extracts all of the txns in it, and includes those, plus all mempool txns into their block. Here we assume none of the blocks are full and all txns can fit.

In both SSFR and BPI, assume the Yoink Attack miner has an identical chance of their rollback becoming included in the canonical chain (e.g. they have identical proportional hash power, …). For simplicity, let's assume they have an $H \in [0, 1)$ chance of mining any block, and if they mine a rollback block, the chance it makes it into the canonical chain is $1/2$ whereas if they mine atop the tip the chance it makes it into the canonical chain is $1$.

Now let's calculate expected pay-offs / opportunity costs:

**BPI:**
- **Yoink Attack**: The Yoink Block contains $2X$ in fees and has a chance of success of $\frac{H}{2}$ for an expected value (e.v.) of $H X$.
- **Honest Mining**: Their alternative is to mine on top of the tip for a block paying $X$ with an expected value of $H X$. Therefore in this scenario, there's no incentive to prefer a Yoink attack or to mine honestly.

**SSFR:**
- **Yoink**: The Yoink Block contains $(1-p)X + 2pX$ in ZSF disbursement plus immediate fees, so the e.v. is $\frac{H}{2} ((1-p)X + 2pX) = \frac{H}{2} (X + pX) = \frac{HX}{2} (1 + p)$.
- **Honest**: Their alternative to mine honestly is $H X$.

Comparing the ratio of Yoink / Honest tells us about the incentive to attack where >1 is an incentive to attack and <1 is a disincentive:

$$
\frac{HX}{2} (1 + p) / H X = \frac{ 1 + p }{ 2 }
$$

The case where $p = 1$ is literally BPI, and if $p = 0$ we see the disincentive is 50%: in other words, miners should expect to lose 50% of their revenue by attempting this attack. If $p=0.4$ (as the ticket title suggests), then they should expect to lose 30% of their revenue.

So we see that while **BPI** does not provide any disincentive to a Yoink attack (and no incentive), **SSFR** includes a _disincentive_ to perform this attack in this scenario, and the disincentive is larger the smaller $p$ is.

### Txn Yoink Rollback Attack (Single Block), Big Fee Lump

Ok, that scenario had constant fees per block, but the concern is if the latest block has very large fees, the incentive to Yoink Attack can become dangerous.

Let's model that by assuming every block *except the latest* has $X$ fees, and the latest has $m X$ fees, where $m ≥ 1$ is some multiplier. Also, all blocks following the "Big Fee Lump" block still have $X$ fees.

Finally, to be thorough, let's include the fact that the large deposit to ZSF influences disbursement:

Recall that up to the big fee lump block, we were in SSFR defined as: $(1-p)X$ = `BlockSubsidy(h)`

Let's introduce some definitions:

- $f$ is the `BLOCK_SUBSIDY_FRACTION` from #706 (mnemonic: $f$ for "fraction")
- $B(h)$ is the ZSF balance at height $h$ (mnemonic: $B$ for "balance")
- $W(h)$ is the amount withdrawn from ZSF and paid to a miner at block $h$ (mnemonic: $W$ for "withdrawal") 

Now the SSFR condition can be written as:

$$
W(h) = f B(h)
$$

And for all blocks heights except the big fee lump block we have:

$$
W(h) = (1-p)X
$$

This implies for SSFR that the balance is held constant because $W(h) = (1-p) X = f B(h)$ so:

$$
B(h) = \frac{1-p}{f} X
$$

When the fee lump block is mined, it disturbs this balance by depositing more than the issuance amount. This makes the new balance:

$$
B(h+1) = \frac{1-p}{f} X + (1-p) m X
$$

This means the ZSF issuance at height $h+1$ will now be $W(h+1) = f B(h+1)$, so:

$$
W(h+1) = (1-p) X + f (1-p) m X
$$

With that in mind, let's examine the opportunity cost of attacks:

**BPI:**
- **Yoink**: $\frac{H}{2} (X + m X) = \frac{H}{2} (1 + m) X$
- **Honest**: $H X$

The incentive to attack as a ratio is: $\frac{ \frac{H}{2} (1 + m) X }{ H X } = \frac{ 1 + m }{ 2 }$

**SSFR.**
- **Yoink**: $\frac{H}{2} ((1-p)X + pX + mpX) = \frac{H X}{2} (1 + mp)$
- **Honest**: $H ( W(h+1) + pX ) = H X (1 + f(1-p)mX)$
  - Algebra steps:
    - $H [ W(h+1) + pX ]$
    - $= H [ ( (1-p) X + f (1-p) m X ) + pX ]$
    - $= H [ X + f (1-p) m X ]$
    - $= H X (1 + f(1-p)mX)$

So the attack incentive ratio becomes:

$$
\frac{ 1 + mp }{ 2 + 2f(1-p)mX }
$$

Ok, this is too complex. Since $f$ is so tiny that it's basically 0, let's simplify out that term:

$$
\frac{ 1 + mp }{2}
$$

When $p = 1$ it is equivalent to the BPI case (as seen above in the BPI attack incentive ratio).

When $p = 0$, we see that the ratio is $\frac{1}{2}$, and this is the same ratio as without the big fee lump, which makes sense because no fees go to miners, so the miner loses potential ZSF withdrawal revenue (since we assume their chance of attack is half their chance of mining honestly).

When is the attack disincentivized, i.e. the ratio is 1 or less?

- $\frac{ 1 + mp }{2} \le 1$
- $1 + mp \le 2$
- $mp \le 1$
- $m \le \frac{1}{p}$

This tells us that if $p$ were, say 0.4 (as the ticket title suggests), that a big fee lump block paying $\le 2.5 X$ would not have produce an incentive for a Yoink rollback, whereas a lump of $\gt 2.5 X$ would.

## Security Brainstorm Conclusion

For the "Steady Fee Rate Regime", which Zcash must pass through prior to reaching the 0-issuance regime, this ZIP along with #706 *disincentivizes* Yoinking attacks more so than Bitcoin-like design. There can still be an incentive to perform yoinking attacks for large fee blocks (but it is relatively less than the Bitcoin equivalent).

Reply:
Thanks @nathan-at-least and thanks for the edit :)

I have added the tautology to the ZIP

Reply:
It might also be useful to survey whether other cryptocurrencies have made any roughly similar changes in issuance or fees, and what the effects on them have been.

Reply:
> If median transaction fees are small, then miners have a smaller incentive to continuously mine many transactions. So a smaller ZSF deposit fraction would be more secure.

I see what you mean, but the second sentence is ambiguous. I think it is intended to mean: "This dilutes the negative impact on security of a smaller ZSF deposit fraction." (I.e. a smaller ZSF deposit fraction is still less secure than a larger one wrt rollback incentives, just not by as much.)